商品購入機能

kotaro0809 opened the pull request:
# 商品購入機能

# 商品購入機能実装の為



必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/2771af6d57685c5b64a6037fbc456331

入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/4b2f9f5fcfcebb690f9b55d61730c4c6

 ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/e48324635be30a45cdf48d6b455430ec
https://gyazo.com/979ae9510567676d0b9b025a5f471ebb

ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/97b8455de4bdeac366a5a47e655dc2d7
https://gyazo.com/50260974be3eeac958aeb7605dad3360

ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/74b9b86238f8c11812b129b3ed753743

売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/64f6a180b12bb59b084b87298f57ce96
https://gyazo.com/c16a3e5ab934d8da77a22ff6ed77f491

売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/a9f85832fa2020458a5fe9e33f0e0a4b

ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/5f9e2f62ba912dfde684a9a91b3bf1b5

ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/39e5d91e305de77c2774dc11c0ac64ef


テスト結果の画像
https://gyazo.com/8c2c680253377611366402ed63e5f27b

kotaro0809 committed:
rspec

kotaro0809 committed:
編集

kotaro0809 committed:
index createアクション定義

kotaro0809 committed:
fromオブジェクトファイル作成

kotaro0809 committed:
アソシエーション定義

kotaro0809 committed:
購入ボタンの遷移先設定

item_purchase_records_path(@item.id)の設定

kotaro0809 committed:
購入記録デーブル作成

マイグレート

kotaro0809 committed:
ルーティング追加

purchase_records create ondex

kotaro0809 committed:
hoge → 配送先の情報に変更

kotaro0809 committed:
hoge → 配送先のカラム名に変更

kotaro0809 committed:
controller編集

privateメソッド
before_action :move_to_root, only: [:index]

未ログイン購入rootへ遷移
 before_action :authenticate_user!

その他保存の記述

kotaro0809 committed:
view編集

商品名　送料　価格

kotaro0809 committed:
fromオブジェクト編集

:purchase_record_idはバリデーションとattr_accessorには記述しない

kotaro0809 committed:
item/view/show編集

kotaro0809 committed:
商品購入jsファイル初期設定

kotaro0809 committed:
pay.jp読み込み

 <script type="text/javascript" src="https://js.pay.jp/v1/"></script>

kotaro0809 committed:
index.erb編集

hoge→カード情報

kotaro0809 committed:
トークン取得まで実装

kotaro0809 committed:
card.jsからフォーム送信まで実装

kotaro0809 committed:
pay.js導入

kotaro0809 committed:
purchase_records_destination.rb編集

tokenの扱いを許可
とバリデーションの設定

kotaro0809 committed:
sold out 表示

kotaro0809 committed:
index.erb編集

売却済み商品編集rootに遷移

kotaro0809 committed:
決済情報送信　環境変数設定

kotaro0809 committed:
purchase_recordsコントローラー編集

決済　購入済み商品購入不可の処理

kotaro0809 committed:
itemsコントローラー編集

購入済み商品rootに遷移

kotaro0809 committed:
sold out 表示

kotaro0809 committed:
環境変数設定

kotaro0809 committed:
商品購入機能モデル単体テスト

Mysql2::Error: MySQL client is not connected
テストコード実行時に上記のエラーが出る
容量オーバーのため？
インスタンス生成後に　sleep 0.5
を記述し処理間の猶予を与えたら通る

kotaro0809 committed:
index.erb

商品購入ページに商品画像表示

kotaro0809 commented:
必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/a59eb0f2f81380deb863955caefa18c3
https://gyazo.com/6517d5b9968d580e8b5f715449c25173

入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/bb2144ccc706f82103534d8ccfa2ab75

 ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/6594ff512ed669e84eebaad56318502b
https://gyazo.com/6594ff512ed669e84eebaad56318502b

ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/31bf07882a4e2f3bf8aa3d39040b724b
https://gyazo.com/323d015270111f6b3967572b3ec54e09

ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/74b9b86238f8c11812b129b3ed753743

売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/5f176aefe02c2bfed3bd117b248f724c
https://gyazo.com/5f176aefe02c2bfed3bd117b248f724c
https://gyazo.com/9041c07d1757307c346d698aa6c2db3f


ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/5f9e2f62ba912dfde684a9a91b3bf1b5

ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/39e5d91e305de77c2774dc11c0ac64ef


テスト結果の画像
https://gyazo.com/8c2c680253377611366402ed63e5f27b


修正しました
ご確認お願いします！

zonodance reviewed (commented):
レビュー依頼ありがとうございます！
下記動画がタイトルとの内容に誤りがあります。
・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/bb2144ccc706f82103534d8ccfa2ab75

上記タイトルに沿った動画を提出の上、再度プルリクエストのご提出をお願いいたします。

kotaro0809 commented:
必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/a59eb0f2f81380deb863955caefa18c3
https://gyazo.com/6517d5b9968d580e8b5f715449c25173

入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/51226d46274c98d545b9c8ba23df3eff
※更新しました

ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/6594ff512ed669e84eebaad56318502b
https://gyazo.com/6594ff512ed669e84eebaad56318502b

ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/31bf07882a4e2f3bf8aa3d39040b724b
https://gyazo.com/323d015270111f6b3967572b3ec54e09

ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/74b9b86238f8c11812b129b3ed753743

売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/5f176aefe02c2bfed3bd117b248f724c
https://gyazo.com/5f176aefe02c2bfed3bd117b248f724c
https://gyazo.com/9041c07d1757307c346d698aa6c2db3f

ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/5f9e2f62ba912dfde684a9a91b3bf1b5

ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/39e5d91e305de77c2774dc11c0ac64ef

テスト結果の画像
https://gyazo.com/8c2c680253377611366402ed63e5f27b

修正しました
ご確認お願いします！

div-reviewer reviewed (commented):
kotaro0809様、学習お疲れ様です！
お忙しい中商品購入の再レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

動画の修正もありがとうございました。一緒に確認させて頂きました。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

kotaro0809 committed:
presence: true　重複削除

kotaro0809 committed:
phon_number のバリデーション編集

10桁でも可能に変更

kotaro0809 committed:
一部変更

 user_id { Faker::Number.non_zero_digit }
    item_id { Faker::Number.non_zero_digit }
を削除
before Actionで生成されている為不要

kotaro0809 committed:
テスト項目追加

・user_id（購入者）が空だと購入できない
・item_id（購入商品）が空だと購入できない

kotaro0809 committed:
phon_numberのテスト項目追加

・電話番号が9桁以下だと登録できない
・電話番号が12桁以上だと登録できない

kotaro0809 commented:
確認しました！

テスト項目を追加しテストに成功する画像です
https://gyazo.com/24c73decb8007e3cf76b7f4778b474a0

よろしくお願いします！

div-reviewer reviewed (commented):
kotaro0809様、学習お疲れさまです。
お忙しい中商品購入の再レビュー依頼ありがとうございました。

前回のコメントにあった修正点が綺麗に直せましたね。素晴らしいです。
こちら`LGTM`とさせて頂きます。実装お疲れさまでした。

**●任意修正について●**
タイトルにあります通り、修正は任意です。
よりよいコードを書くための知識としてご確認いただけますと幸いです。

修正して頂ける場合は`再度レビュー依頼は出さなくて大丈夫です`ので、修正後、コミットとプッシュして頂けましたら、
マージ後、ローカルのmasterブランチに忘れずにプルし、次の実装に取り組んで頂ければと思います。

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
今後の学習も応援しております！どうぞよろしくおねがいします。

kotaro0809 committed:
リファンダリングのため修正

kotaro0809 committed:
rubocop　修正
